streams/src/main/java/org/apache/kafka/streams/errors/ErrorHandlerContext.java

```diff
@@ -102,4 +106,23 @@ public interface ErrorHandlerContext {
      * @return the task ID
      */
     TaskId taskId();
+
+    /**
+     * Return the current timestamp.
+     *
+     * <p> If it is triggered while processing a record streamed from the source processor,
+     * timestamp is defined as the timestamp of the current input record; the timestamp is extracted from
+     * {@link org.apache.kafka.clients.consumer.ConsumerRecord ConsumerRecord} by {@link TimestampExtractor}.
+     * Note, that an upstream {@link org.apache.kafka.streams.processor.api.Processor} might have set a new timestamp by calling
+     * {@link ProcessorContext#forward(Object, Object, To) forward(..., To.all().withTimestamp(...))}.
+     * In particular, some Kafka Streams DSL operators set result record timestamps explicitly,
+     * to guarantee deterministic results.
+     *
+     * <p> If it is triggered while processing a record generated not from the source processor (for example,
+     * if this method is invoked from the punctuate call), timestamp is defined as the current
+     * task's stream time, which is defined as the largest timestamp of any record processed by the task.
```

Comment:
Is this correct for all cases? It's for sure correct for the old PAPI using `processor.ProcessorContext`. But I am not sure about the new PAPI which uses `processor.api.ProcessorContext` for which `forward(Record)` take a `Record` which will have a timestamp?

Can we verify this?

Reply:
I  have updated th comment with the new processor.api.ProcessorContext

Reply:
(Did you verify the behavior?)

Also, I don't see an update?

Last, given that `ErrorHandlerContext` in also used for the deserialization case, I am wondering if we should say something about this, too? For this case, the `TimestampExtractor` was not called yet, and thus we pass in `rawRecord.timestamp()`.

Reply:
I did some more digging... Seems, we actually need to distinguish between `stream time` vs `wall-clock time` punctuations.

For `stream time` punctuation, we set "stream time" as "record context timestamp" and thus `context.timestamp()` will return "stream time". For `wall-clock time` punctuations, we set "system time" as "record context timestamp" and thus `context.timestamp()` will return "system time".

Did I get this correctly?

Reply:
Did you copy this from `RecordContext#timestamp` JavaDocs? They might also need an update covering the wall-clock time case?

Overall, it seems there is no difference between old and new PAPI though?

Reply:
> (Did you verify the behavior? I would hope that TTD doe the right/same thing as the actual runtime for this case.)
> 
> Last, given that `ErrorHandlerContext` in also used for the deserialization case, I am wondering if we should say something about this, too? For this case, the `TimestampExtractor` was not called yet, and thus we pass in `rawRecord.timestamp()`.

I added a comment when it comes from a deserialization failure

Reply:
> I did some more digging... Seems, we actually need to distinguish between `stream time` vs `wall-clock time` punctuations.
> 
> For `stream time` punctuation, we set "stream time" as "record context timestamp" and thus `context.timestamp()` will return "stream time". For `wall-clock time` punctuations, we set "system time" as "record context timestamp" and thus `context.timestamp()` will return "system time".
> 
> Did I get this correctly?

On what I saw, you are right, I have updated the java doc by splitting different cases of puntuation type

Reply:
> Did you copy this from `RecordContext#timestamp` JavaDocs? They might also need an update covering the wall-clock time case?
> 
> Overall, it seems there is no difference between old and new PAPI though?

Indeed, I have copied the java doc. 

If comments on the ErrorHandlerContext are good for you, do I have to change it with the new API and add the wall clock time part ?